src/runtime/HalideBuffer.h

```diff
@@ -167,20 +181,34 @@ class Buffer {
         if (alloc) {
             int new_count = --(alloc->ref_count);
             if (new_count == 0) {
-                if (buf.dev) {
-                    halide_device_free_t fn = halide_get_device_free_fn();
-                    assert(fn &&
-                           "Buffer has a device allocation but no Halide Runtime linked");
-                    (*fn)(nullptr, &buf);
-                }
                 void (*fn)(void *) = alloc->deallocate_fn;
                 fn(alloc);
             }
+            drop_device_allocation();
             alloc = nullptr;
             buf.host = nullptr;
         }
     }
 
+    void drop_device_allocation() {
```

Comment:
There doesn't seem to be a way to get at this functionality from the public API. (Short of doing a nop crop or slice operation.) It seems this would be useful to explicitly limit scope lifetime more safely than calling device_free in some instances.